docs/proposals/20220927-label-sync-between-machine-and-nodes.md

```diff
@@ -0,0 +1,131 @@
+
+- The `node-restriction.kubernetes.io/` prefix is recommended in the [Kubernetes docs](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for things such as workload placement; according to https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/ also
+  labels like `example.node-restriction.kubernetes.io/fips=true` falls in this category (prefix =~ domain).
+
+- Cluster API will define its own label prefix, `node.cluster.x-k8s.io/`; for consistency we consider also `example.node.cluster.x-k8s.io/gpu=true` in this category (prefix =~ domain).
+
+#### Synchronization of CAPI Labels
+
+Only the labels that fall under the prefix described above will need to be kept in sync between the Machine and Node objects.
+
+The syncronization process is going to use [server side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) in order to ensure that Cluster API will manage only the subset of labels coming from Machine objects and ignore labels applied concurrently by other users/other components.
+
+This requires to define an identity/manager name to be used by CAPI when performing this operation; additionally during implementation we are going to identify and address eventual additional steps required to properly transition existing Nodes to SSA, if required.
+
+The syncronization process will be implemented in the Machine controller, so reconciliation is triggered both when Machine object changes or when the Node changes (the machine controller watches Nodes within the workload cluster).
+
+#### Delay between Node Create and Label Sync
+
+The Node object is first created when kubeadm joins a node to the workload cluster (i.e. kubelet is up and running). There may be a delay (potentially several seconds) before the machine controller kicks in to apply the labels on the Node.
+
+Kubernetes supports both equality and inequality requirements in label selection. In an equality based selection, the user wants to place a workload on node(s) matching a specific label (e.g. Node.Labels contains `my.prefix/foo=bar`). The delay in applying the label on the node, may cause a subsequent delay in the placement of the workload, but this is likely acceptable.
+
+In an inequality based selection, the user wants to place a workload on node(s) that do not contain a specific label (e.g. Node.Labels not contain `my.prefix/foo=bar`). The case is potentially problematic because it relies on the absence of a label and this can occur if the pod scheduler runs during the delay interval.
+
+One way to address this is to use kubelet's `--register-with-taints` flag. Newly minted nodes can be tainted via the taint `cluster.x-k8s.io=label-sync-pending:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints).
```

Comment:
```suggestion
One way to address this is to use kubelet's `--register-with-taints` flag. Newly minted nodes can be tainted via the taint `node.cluster.x-k8s.io=machine-pending:NoSchedule`. Assuming workloads don't have this specific toleration, then nothing should be scheduled. KubeadmConfigTemplate provides the means to set taints on nodes (see  JoinConfiguration.NodeRegistration.Taints).
```

It might be useful to define the taint in the newly defined domain, and make it generic enough that it can be used for other initialization steps as well in the future.

Reply:
what about `node.cluster.x-k8s.io=initialization-pending:NoSchedule`

Reply:
Given

> kubelets specifying --cloud-provider=external will add a taint node.cloudprovider.kubernetes.io/uninitialized with an effect NoSchedule during initialization. This marks the node as needing a second initialization from an external controller before it can be scheduled work. Note that in the event that cloud controller manager is not available, new nodes in the cluster will be left unschedulable. The taint is important since the scheduler may require cloud specific information about nodes such as their region or type (high cpu, gpu, high memory, spot instance, etc).

What about: `node.cluster.x-k8s.io=uninitialized:NoSchedule`

Reply:
updated to node.cluster.x-k8s.io=uninitialized:NoSchedule